modules/packages/arch/metadata.go

```diff
@@ -0,0 +1,330 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package arch
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"crypto/md5"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/mholt/archiver/v3"
+)
+
+// Metadata for arch package.
+type Metadata struct {
+	Filename          string
+	Name              string
+	Base              string
+	Version           string
+	Description       string
+	CompressedSize    int64
+	CompressedSizeMib string
+	InstalledSize     int64
+	InstalledSizeMib  string
+	MD5               string
+	SHA256            string
+	URL               string
+	BuildDate         int64
+	BuildDateStr      string
+	BaseDomain        string
+	Packager          string
+	Provides          []string
+	License           []string
+	Arch              []string
+	Depends           []string
+	OptDepends        []string
+	MakeDepends       []string
+	CheckDepends      []string
+	Backup            []string
+}
+
+// Function that recieves arch package archive data and returns it's metadata.
+func EjectMetadata(filename, domain string, pkg []byte) (*Metadata, error) {
+	pkgreader := io.LimitReader(bytes.NewReader(pkg), 250000)
+	var buf bytes.Buffer
+	err := archiver.DefaultZstd.Decompress(pkgreader, &buf)
+	if err != nil {
+		if !errors.Is(err, io.ErrUnexpectedEOF) {
+			return nil, err
+		}
+	}
+	splt := strings.Split(buf.String(), "PKGINFO")
+	if len(splt) < 2 {
+		return nil, errors.New("unable to eject .PKGINFO from archive")
+	}
+	raw := splt[1][0:10000]
```

Comment:
This looks really fragile. The package is just a compressed tar archive, so unpack it and process the `PKINFO` file without reinpreting everything as a string.